addons/website_sale_wallet/models/sale_order.py

```diff
@@ -0,0 +1,110 @@
+    def _compute_gift_card_purchased(self):
+        self.generated_gift_card_count = len(self.generated_gift_card_ids)
+
+    # Logic
+    def pay_using_wallet_partner_in_website(self, partner, website):
+        can_be_retained = partner.how_much_can_be_retained(self.amount_total)
+        if can_be_retained > 0:
+            product_gift_card_pay = self._get_wallet_product()
+            if self.order_line_wallet_id:
+                self.order_line_wallet_id.price_unit = - can_be_retained
+            else:
+                self.order_line_wallet_id = self.env["sale.order.line"].create({
+                    'product_id': product_gift_card_pay.id,
+                    'price_unit': - can_be_retained,
+                    'product_uom_qty': 1,
+                    'product_uom': product_gift_card_pay.uom_id.id,
+                    'wallet_id': partner.id,
+                    'website_id': website.id,
+                    'order_id': self.id
+                })
+            return True
+        return False
+
+    def _get_wallet_product(self):
+        if not self._wallet_product_id:
+            product = self.env['product.product'].search([('default_code', '=', 'PAY_WITH_WALLET')])
+            if product:
+                self._wallet_product_id = product
+            else:
+                self._wallet_product_id = self.env['product.product'].create({
+                    'default_code': 'PAY_WITH_WALLET',
+                    'name': _("Pay with Wallet"),
+                    'type': 'service',
+                    'taxes_id': False,
+                    'supplier_taxes_id': False,
+                    'sale_ok': False,
+                    'purchase_ok': False,
+                    'lst_price': 0,
+                })
+        return self._wallet_product_id
+
+    def _create_gift_cards(self, gift_card_order_line):
+        return [self.env['gift.card'].create({
+            'initial_balance': gift_card_order_line.price_unit,
+            'sale_order_purchase_id': self.id,
+            'website_id': self.env['website'].get_current_website().id
+        }) for _ in range(int(gift_card_order_line.product_uom_qty))]
+
+    def _send_gift_card_mail(self):
+        template = self.env.ref('gift_card_template_email', raise_if_not_found=False)
+        if template:
+            for order in self:
+                for gift_card in order.generated_gift_card_ids:
+                    order.message_post_with_template(
+                        template.id, composition_mode='comment',
+                        model='gift.card', res_id=gift_card.id,
+                        email_layout_xmlid='mail.mail_notification_light',
+                    )
+
+
+class SaleOrderLine(models.Model):
+    # Inherited Fields
+    _inherit = "sale.order.line"
+
+    # Relational fields
+    wallet_id = fields.Many2one("res.partner", default=None)
+    website_id = fields.Many2one("website", ondelete="restrict", string="Website")
```

Comment:
does it make sense to have this field ? 

Order with
L1 10€
L2 15€
L3 5€

You apply a wallet of 17€ on which line ? 

Imo, these fields should no be on sale order line.
It you want the wallet_id from the oder, you need a related field.

Reply:
wallet_id will be available only for the new order_line created (pay with wallet)

Reply:
probably useless because we know that the product_id  is wallet no?

Reply:
I'm trying to remove all of these fields and add a boolean field instead (is_wallet_payment)